### PR TITLE
Add ToolApprovalRequestContent.RequiresConfirmation

### LINKED ISSUE #7550: [Microsoft.Extensions.AI] Add ToolApprovalRequestContent.IsInvokerRequested
## Summary

`Microsoft.Extensions.AI.ToolApprovalRequestContent` gains a new `IsInvokerRequested` property so consumers (UIs, policy layers, telemetry) can distinguish approval requests that exist because the underlying function is an `ApprovalRequiredAIFunction` from approval requests that an invoker (typically `FunctionInvokingChatClient`, "FICC") added as collateral because a peer call in the same response required approval. This is the only information needed to render an approval prompt **only** for the calls that genuinely demanded one, while silently auto-approving the collateral peers — a workflow that has no general solution today once FICC is wrapped behind additional `IChatClient` layers.

Draft implementation: #7549.

## Motivation

`FunctionInvokingChatClient` enforces an important invariant: if any function call in a single assistant response targets an `ApprovalRequiredAIFunction`, **all** function calls in that response are converted into `ToolApprovalRequestContent` items before being yielded to the next layer. This is required so approvals and rejections stay coherent across the batch — the client cannot partially execute some calls while waiting on approval for others without losing the ability to correctly attribute results.

The consequence is that downstream consumers (in particular UI layers that render approval prompts) receive a mixed bag of `ToolApprovalRequestContent` items where some correspond to functions the application explicitly marked as approval-required and others are pure collateral. Today there is no way to tell which is which from the content itself.

In principle a consumer could re-resolve the tool by name against `ChatOptions.Tools`/`AdditionalTools` and check `tool.GetService<ApprovalRequiredAIFunction>()`, but in practice this does not work for any non-trivial pipeline:

- FICC is commonly wrapped by additional `DelegatingChatClient` layers (telemetry, caching, evaluation, OpenTelemetry, retry, governance, an `IChatClient` exposed by an agent framework, etc.). The UI sits at the top of the pipeline and has no guaranteed handle on the specific tool list FICC saw.
- The tool list can be mutated by tools themselves during the invocation loop, so what the UI inspects later may not match what FICC saw at the moment of wrapping.
- Function names are not guaranteed unique, and the matching logic (which mirrors FICC's own `FindTool`) would have to be reimplemented and kept in sync.
- For UI scenarios that hydrate `ChatMessage` instances from persistence, the originating tools may not be available at all.

FICC, by contrast, **does** know the answer at the exact moment it wraps each call. Exposing that decision on the resulting `ToolApprovalRequestContent` lets every downstream consumer behave correctly without re-implementing FICC's resolution logic.

## Goals

- Make it trivial for any consumer of `ToolApprovalRequestContent` — UI components, policy layers, loggers, evaluators — to tell, per request, whether the underlying tool itself required approval or the request was wrapped as collateral by an invoker.
- Enable UIs to prompt the user only for genuinely approval-required calls and auto-approve the collateral ones, without inspecting `ChatOptions.Tools` or `AdditionalTools`.
- Preserve the existing batch-coherence behavior of `FunctionInvokingChatClient`: approvals and rejections in a response still flow as a single coordinated batch; the new signal is purely informational.
- Work uniformly across the non-streaming and streaming code paths.
- Survive serialization, so the distinction is preserved when conversation history is persisted and rehydrated.
- Remain robust when an `ApprovalRequiredAIFunction` is nested inside additional `DelegatingAIFunction` wrappers.

## Non-goals

- **Not a new declarative mechanism for marking functions as approval-required.** `ApprovalRequiredAIFunction` remains the way to opt a function in to approval.
- **Not a change in batching behavior.** FICC continues to wrap every concurrent function call in a response when any one of them requires approval. The new property only labels each resulting request; it does not let some calls in a batch be executed while others wait.
- **Not a per-call dynamic approval predicate.** The property describes a static fact about the tool the call targets at the moment of wrapping, not "should this specific arguments combination require approval".
- **Not a change to `FunctionCallContent`.** The asymmetry between "tool required approval" and "invoker wrapped a peer" is a property of the approval request envelope, not the underlying call.
- **Not a generalized `ToolApprovalRequestSource` taxonomy.** A future enum (policy-based, MCP-server-declared, etc.) could subsume the boolean, but introducing one is out of scope for this proposal.

### New API

`Microsoft.Extensions.AI.Abstractions` adds a single member to the existing `ToolApprovalRequestContent` class:

```csharp
public sealed class ToolApprovalRequestContent : InputRequestContent
{
    // ... existing members unchanged ...

    /// <summary>
    /// Gets or sets a value indicating whether this approval request was added by the invoker
    /// of the tool call rather than because the underlying tool itself was declared as
    /// requiring approval.
    /// </summary>
    /// <remarks>
    /// Defaults to <see langword="false"/>, meaning the underlying tool itself required
    /// approval (for example, the targeted function was an <see cref="ApprovalRequiredAIFunction"/>).
    /// Set to <see langword="true"/> when the invoker (such as
    /// <see cref="FunctionInvokingChatClient"/>) wrapped a peer call as collateral so that
    /// approvals and rejections stay coherent across a batch in which at least one peer
    /// genuinely required approval.
    /// </remarks>
    public bool IsInvokerRequested { get; set; }
}
```

The property is a mutable, serialized boolean — same shape as `FunctionCallContent.InformationalOnly` — and is introduced under the existing `MEAI001` experimental diagnostic that already covers the approval feature.

### Polarity and default

- `false` (default) means "the targeted function itself requires approval". This is the safe default for any non-FICC producer that constructs a `ToolApprovalRequestContent` directly because the tool genuinely demands approval — the property is correct without any code change at the construction site.
- `true` means "an invoker added this request as collateral". The targeted function did not itself require approval; the request exists only to keep the batch coherent. Consumers may treat such requests as candidates for auto-approval.

The affirmative direction matches the role FICC plays in the pipeline: the request was *invoker-requested*, as opposed to *function-requested*.

### Producer behavior (`FunctionInvokingChatClient`)

FICC sets the flag at both points where it converts `FunctionCallContent` into `ToolApprovalRequestContent`:

1. The non-streaming response path, where FICC walks the assembled messages and rewrites function calls into approval requests when any peer requires approval.
2. The streaming response path, where FICC buffers updates until it knows whether the batch contains an approval-required call and then flushes a corrected sequence of updates.

Both paths use the same per-call resolution that FICC already performs to drive batch wrapping: look up the function by name across `ChatOptions.Tools` and `AdditionalTools`, then ask `tool.GetService<ApprovalRequiredAIFunction>()`. Using `GetService` rather than a direct type test is important because it walks `DelegatingAIFunction` chains; a tool declared as `Telemetry(Logging(ApprovalRequired(realFn)))` must still be classified as approval-required.

`IsInvokerRequested` is set to `true` exactly when this resolution returns `null` for the call being wrapped — i.e. when no `ApprovalRequiredAIFunction` is visible behind the targeted function. Otherwise it remains `false`.

### Consumer behavior

Consumers that want today's behavior (treat every `ToolApprovalRequestContent` as something the user must approve) need no changes. Consumers that want to opt into the new behavior — auto-approving collateral requests while only prompting for the genuine ones — can do so based on the flag alone:

```csharp
foreach (var request in message.Contents.OfType<ToolApprovalRequestContent>())
{
    if (request.IsInvokerRequested)
    {
        // Collateral: auto-approve without prompting.
        responses.Add(request.CreateResponse(approved: true));
    }
    else
    {
        // Genuine: surface the approval prompt in the UI.
        pendingPrompts.Add(request);
    }
}
```

This works regardless of how many `IChatClient` layers sit between the consumer and FICC, and without the consumer needing to inspect, replicate, or even have access to `ChatOptions.Tools` or `AdditionalTools`.

### Persistence

`IsInvokerRequested` is a regular serialized property: it round-trips through `System.Text.Json` using `AIJsonUtilities.DefaultOptions` the same way the rest of `ToolApprovalRequestContent` does. Conversation histories that include approval requests preserve the distinction across persist/rehydrate cycles, so UIs that re-render an existing thread continue to make the right decision for previously-recorded requests.

### Interaction with non-function tool calls

FICC only constructs `ToolApprovalRequestContent` for `FunctionCallContent`. Approval requests for other tool call kinds (for example MCP server tool calls produced by an upstream service) are passed through unchanged; their `IsInvokerRequested` retains whatever value the producer set, which by default is `false`. This matches the polarity contract: a `ToolApprovalRequestContent` that arrives from outside FICC is, by definition, not invoker-added by FICC.

### PR DESCRIPTION
**Pull request by javiercn**
Resolves #7550.

Adds `ToolApprovalRequestContent.RequiresConfirmation` so consumers can distinguish approval requests where the underlying function genuinely needs approval (`RequiresConfirmation = true`, the default) from requests that `FunctionInvokingChatClient` added as a no-op because a peer call in the same response requires approval (`RequiresConfirmation = false`).

See #7550 for the full proposal (summary, motivation, goals, non-goals, detailed design). The name is still being iterated on in #7550; the property name in this PR may change to the agreed-upon name before merge.

## Changes

- **`Microsoft.Extensions.AI.Abstractions`** — new `bool ToolApprovalRequestContent.RequiresConfirmation { get; set; } = true;`, serialized like `FunctionCallContent.InformationalOnly`. Default `true` means any non-FICC producer that constructs a `ToolApprovalRequestContent` because the underlying tool genuinely requires approval is correct without code changes.
- **`FunctionInvokingChatClient`** — both wrap sites now set the flag per call.
  - Non-streaming `ReplaceFunctionCallsWithApprovalRequests` uses `FindTool` + `GetService<ApprovalRequiredAIFunction>()` per FCC; the lookup is captured in the existing index list so the wrap site sets `RequiresConfirmation` without a second pass.
  - Streaming `TryReplaceFunctionCallsWithApprovalRequests` now takes the pre-computed `approvalRequiredFunctions` array (already computed at the caller for batch buffering) and matches by name, avoiding a second `FindTool` + `GetService` scan per FCC.
- **API baseline** — `Microsoft.Extensions.AI.Abstractions.json` updated.

## Tests

- `AssertExtensions.EqualMessageLists` and the test-local `CloneFcc` propagate and assert the new flag.
- Two existing FICC mixed-batch approval tests assert per-call polarity.
- Two new FICC tests: (a) approval-required function nested inside another `DelegatingAIFunction`; (b) FCC whose name doesn't match any tool when a peer requires approval.
- Two new unit tests on `ToolApprovalRequestContent`: default value, JSON round-trip.

40 approval tests, 161 FICC tests, and 21 `ToolApprovalRequestContent` unit tests pass on net9.0.

### CODE REVIEW

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs [hunk 1]
```diff
@@ -32,6 +34,27 @@ public ToolApprovalRequestContent(string requestId, ToolCallContent toolCall)
...
+    /// <para>
+    /// Some invokers convert every concurrent function call in a response into a
+    /// <see cref="ToolApprovalRequestContent"/> whenever any one of them targets an
+    /// <see cref="ApprovalRequiredAIFunction"/>, so that approvals and rejections stay coherent
+    /// across the batch. When set to <see langword="true"/>, this property indicates that the
+    /// request was added for that reason and the underlying tool did not itself require approval;
+    /// consumers may use this to, for example, auto-approve such requests.
+    /// </para>
+    /// </remarks>
+    [Experimental(DiagnosticIds.Experiments.AIFunctionApprovals, UrlFormat = DiagnosticIds.UrlFormat)]
```
Comment (jozkee) [inline src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: Approval APIs have dropped the `Experimental` attribute.  `AIFunctionApprovals` is a leftover.  I don't mind using it here since in the end the user will see MEAI001.  Just wanted to point it out.

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs [hunk 2]
```diff
@@ -613,7 +613,7 @@ public override async IAsyncEnumerable<ChatResponseUpdate> GetStreamingResponseA
                     for (; lastYieldedUpdateIndex < updates.Count; lastYieldedUpdateIndex++)
                     {
                         var updateToYield = updates[lastYieldedUpdateIndex];
-                        if (TryReplaceFunctionCallsWithApprovalRequests(updateToYield.Contents, out var updatedContents))
+                        if (TryReplaceFunctionCallsWithApprovalRequests(updateToYield.Contents, out var updatedContents, options?.Tools, AdditionalTools))
```
Comment (jozkee) [inline src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs]: Pass `approvalRequiredFunctions` instead.

```suggestion
                        if (TryReplaceFunctionCallsWithApprovalRequests(updateToYield.Contents, approvalRequiredFunctions, out var updatedContents))
```

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs [hunk 3]
```diff
@@ -1618,7 +1618,10 @@ private static (bool hasApprovalRequiringFcc, int lastApprovalCheckedFCCIndex) C
     /// Replaces all <see cref="FunctionCallContent"/> with <see cref="ToolApprovalRequestContent"/> and ouputs a new list if any of them were replaced.
     /// </summary>
     /// <returns>true if any <see cref="FunctionCallContent"/> was replaced, false otherwise.</returns>
-    private static bool TryReplaceFunctionCallsWithApprovalRequests(IList<AIContent> content, out List<AIContent>? updatedContent)
+    private static bool TryReplaceFunctionCallsWithApprovalRequests(
+        IList<AIContent> content,
+        out List<AIContent>? updatedContent,
+        params ReadOnlySpan<IList<AITool>?> toolLists)
```
Comment (jozkee) [inline src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs]:
```suggestion
        AITool[] approvalRequiredFunctions,
        out List<AIContent>? updatedContent)
```

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs [hunk 4]
```diff
@@ -32,6 +34,27 @@ public ToolApprovalRequestContent(string requestId, ToolCallContent toolCall)
...
+    public bool IsInvokerRequested { get; set; }
```
Comment (jozkee) [inline src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]:
```suggestion
    public bool IsAutoApprovalRecommended { get; set; }
```

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs (hunk 1 above)
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: Good call, dropped in 09a62746da.

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs (hunk 2 above)
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs]: Done in 09a62746da. Helper now matches names against the pre-filtered `approvalRequiredFunctions` array.

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs (hunk 3 above)
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs]: Applied in 09a62746da.

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs (hunk 4 above)
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: I considered the action-recommendation framing (`CanAutoApprove` / `IsAutoApprovable` / `IsAutoApprovalRecommended`) and leaned away from it for a couple of reasons:

- "Recommended" reads as advice the consumer can take or leave, but when FICC sets the value it's making a statement about the underlying tool, not a recommendation.
- I'd rather the property describe what the request is than what the consumer should do with it. That keeps the same surface useful for UI prompts, policy gates, telemetry, and replay without any of them having to reinterpret the wording.

That said, this is mostly a phrasing call and either polarity reads cleanly with the right name. Happy to keep iterating on the name in #7550 so we can settle on a concrete option there before flipping the PR, rather than churning the implementation. Will park this thread and pick up whichever name we agree on.

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs [hunk 5]
```diff
@@ -1629,7 +1632,21 @@ private static bool TryReplaceFunctionCallsWithApprovalRequests(IList<AIContent>
...
+
+                    bool requiresConfirmation = false;
+                    for (int j = 0; j < approvalRequiredFunctions.Length; j++)
+                    {
+                        if (string.Equals(approvalRequiredFunctions[j].Name, fcc.Name, StringComparison.Ordinal))
+                        {
+                            requiresConfirmation = true;
+                            break;
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs:1650]: You're right that the duplicate-name case is undocumented and ambiguous, and FICC already handles it inconsistently across its three name-resolution sites:

- `FindTool` (invocation, line 818) walks `options?.Tools, AdditionalTools` in order and respects precedence — `ChatOptions.Tools` wins.
- `AnyToolsRequireApproval` (line 795) and `CheckForApprovalRequiringFCC` (line 1605, streaming wrap decision) are loose name-set checks across both lists — no precedence.

So a duplicate-named tool already produces inconsistent behavior between invocation and approval gating today, independent of this PR. The only existing test for duplicate names (`RespectsChatOptionsToolsModificationsByFunctionTool_AddToolOverridingAdditionalTool`) covers invocation precedence; no test exercises duplicate names with mixed approval requirements.

What this PR did add was a small streaming/non-streaming divergence for `RequiresConfirmation` itself — non-streaming used `FindTool` while streaming used the loose name-set check. Fixed in 8cc96764c3 by aligning the non-streaming wrap site to take the same precomputed `approvalRequiredFunctions` array and use a name-set check. Both wrap sites now agree.

The deeper question — should approval gating respect tool precedence the way invocation does — is a pre-existing FICC issue that touches `AnyToolsRequireApproval`, `CheckForApprovalRequiringFCC`, and now `RequiresConfirmation` together. Better treated as a separate follow-up so it can be designed consistently across all three sites.

File: src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs [hunk 6]
```diff
@@ -1676,15 +1694,18 @@ private IList<ChatMessage> ReplaceFunctionCallsWithApprovalRequests(
...
             {
                 // Clone the message if we didn't already clone it in a previous iteration.
                 var message = lastMessageIndex != messageIndex ? outputMessages[messageIndex].Clone() : outputMessages[messageIndex];
                 message.Contents = [.. message.Contents];
 
                 var functionCall = (FunctionCallContent)message.Contents[contentIndex];
                 LogFunctionRequiresApproval(functionCall.Name);
-                message.Contents[contentIndex] = new ToolApprovalRequestContent(ComposeApprovalRequestId(functionCall.CallId), functionCall);
+                message.Contents[contentIndex] = new ToolApprovalRequestContent(ComposeApprovalRequestId(functionCall.CallId), functionCall)
+                {
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs:1721]: Going to leave this as-is. The log fires when FICC converts a function call into an approval request, which is still factually true for wraps where `RequiresConfirmation` is false (the wrap happened because a peer in the batch needed approval). It's a diagnostic log at the FICC level rather than user-facing telemetry, and a separate event for the auto-approved case would just add noise.

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs [hunk 7]
```diff
@@ -32,6 +32,18 @@ public ToolApprovalRequestContent(string requestId, ToolCallContent toolCall)
...
+    /// <summary>
+    /// Gets or sets a value indicating whether the underlying tool call must be confirmed
+    /// before it is invoked.
+    /// </summary>
+    /// <remarks>
+    /// Defaults to <see langword="true"/>. When <see langword="true"/>, the underlying tool
+    /// requires a confirmation (such as a user prompt, a policy decision, or any other approver)
+    /// before it can be invoked. When <see langword="false"/>, the underlying tool does not
+    /// require a confirmation and the consumer may proceed without prompting.
+    /// </remarks>
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs:48]: Good point. Clarified in 49394b86e8 that a `ToolApprovalResponseContent` is still required even when `RequiresConfirmation` is false, so the underlying tool call can be invoked.

File: src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs (hunk 1 above)
Comment (jozkee) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: It was not my intent to suggest this should be stable.  I think it should be `[Experimental(DiagnosticIds.Experiments.AIApprovalsInvocationRequired, ...)]` or something along those lines.

Comment (javiercn) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: @copilot chop chop

Comment (jozkee) [reply src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/ToolApprovalRequestContent.cs]: > It was not my intent to suggest this should be stable. I think it should be [Experimental(DiagnosticIds.Experiments.AIApprovalsInvocationRequired, ...)] or something along those lines.

@copilot address my comment.

**Review by jozkee — APPROVED** [review]
LGTM, thanks.

**Review by shyamnamboodiripad — APPROVED** [review]

**Comment by javiercn** [comment]
@jozkee It's been a while since I have done within extensions. What's the process. Do we merge now? Or do we want to go through API Review first or discuss it any further? Is there anyone else that needs to have a look?

**Review by peterwald — APPROVED** [review]

**Comment by jeffhandley** [comment]
@jozkee has inquired with the rest of the .NET AI crew to see if others want to review the API design before we merge. Let's leave it for @jozkee to merge once he's satisfied the crew has weighed in. Thanks for driving this though, @javiercn!
